### PR TITLE
feat(statistics): support for custom date ranges

### PR DESCRIPTION
**Pull request by davwheat**
**Changes proposed in this pull request:**

- add option to choose a custom date range for the statistics widget
- add integration tests for statistics data api requests

**Reviewers should focus on:**

**Screenshot**

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  items(): ItemList<Mithril.Children> {
+    const items = new ItemList<Mithril.Children>();
+
+    items.add('intro', <p>{app.translator.trans('flarum-statistics.admin.date_selection_modal.description')}</p>, 100);
+
+    items.add(
+      'date_start',
+      <div class="Form-group">
+        <label htmlFor={this.state.ids.startDate}>{app.translator.trans('flarum-statistics.admin.date_selection_modal.start_date')}</label>
+        <input type="date" id={this.state.ids.startDate} value={this.state.inputs.startDateVal} onchange={this.updateState('startDateVal')} />
```
Comment (SychO9) [inline extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx]:
```suggestion
        <input type="date" id={this.state.ids.startDate} value={this.state.inputs.startDateVal} onchange={this.updateState('startDateVal')} className="FormControl" />
```

File: extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      </div>,
+      90
+    );
+
+    items.add(
+      'date_end',
+      <div class="Form-group">
+        <label htmlFor={this.state.ids.endDate}>{app.translator.trans('flarum-statistics.admin.date_selection_modal.end_date')}</label>
+        <input type="date" id={this.state.ids.endDate} value={this.state.inputs.endDateVal} onchange={this.updateState('endDateVal')} />
```
Comment (SychO9) [inline extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx]:
```suggestion
        <input type="date" id={this.state.ids.endDate} value={this.state.inputs.endDateVal} onchange={this.updateState('endDateVal')} className="FormControl" />
```

File: extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    if (this.attrs.value) {
+      this.state.inputs = {
+        startDateVal: dayjs.utc(this.attrs.value.start * 1000).format('YYYY-MM-DD'),
+        endDateVal: dayjs.utc(this.attrs.value.end * 1000).format('YYYY-MM-DD'),
+      };
+    }
+  }
+
+  className(): string {
+    return 'StatisticsWidgetDateSelectionModal';
```
Comment (SychO9) [inline extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx]: Just a visual nitpick 😇
```suggestion
    return 'StatisticsWidgetDateSelectionModal Modal--small';
```

File: extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    endDateVal: string;
+  };
+  ids: {
+    startDate: string;
+    endDate: string;
+  };
+}
+
+export default class StatisticsWidgetDateSelectionModal extends Modal<IStatisticsWidgetDateSelectionModalAttrs> {
+  /* @ts-expect-error core typings don't allow us to set the type of the state attr :( */
```
Comment (SychO9) [inline extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx:42]: noticed this too, I guess like we have to always type a default state/attribute type in any core components to not hinder extensibility going forward?

File: extensions/statistics/src/Api/Controller/ShowStatisticsData.php [hunk 5]
```diff
@@ -69,20 +70,36 @@ public function handle(ServerRequestInterface $request): ResponseInterface
...
+        }
+
+        if ($period === 'custom') {
+            if (! $customDateRange) {
+                throw new InvalidParameterException('A custom date range must be specified');
+            }
+
+            // Seconds-based timestamps
+            $startRange = Carbon::createFromTimestampUTC($customDateRange['start'])->toDateTime();
+            $endRange = Carbon::createFromTimestampUTC($customDateRange['end'])->toDateTime();
```
Comment (SychO9) [inline extensions/statistics/src/Api/Controller/ShowStatisticsData.php]: Just to make sure use input is exactly what we expect it to be
```suggestion
        if ($period === 'custom') {
            $start = (int) $customDateRange['start'];
            $end = (int) $customDateRange['end'];

            if (! $customDateRange || ! $start || ! $end) {
                throw new InvalidParameterException('A custom date range must be specified');
            }

            // Seconds-based timestamps
            $startRange = Carbon::createFromTimestampUTC($start)->toDateTime();
            $endRange = Carbon::createFromTimestampUTC($end)->toDateTime();
```

File: extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx (hunk 4 above)
Comment (davwheat) [reply extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx:42]: Yeah, we've done it in some components, but not all.

Just a "simple" `class MyComponent<CustomAttrs extends BaseAttrs = BaseAttrs, CustomState extends BaseState> extends BaseComponent<CustomAttrs, CustomState>` should work.

**Review by SychO9 — APPROVED** [review]

**Pull request merged**